File: packages/agent/src/tools/web/url-validator.ts
/**
 * @fileoverview URL Validator
 *
 * Validates URLs for WebFetch tool with security checks and domain filtering.
 */

import type {
  UrlValidationResult,
  UrlValidatorConfig,
} from './types.js';

const DEFAULT_MAX_LENGTH = 2000;

/**
 * Private IP address patterns
 */
const INTERNAL_PATTERNS = {
  localhost: /^localhost$/i,
  loopbackIPv4: /^127\.\d{1,3}\.\d{1,3}\.\d{1,3}$/,
  zeroAddress: /^0\.0\.0\.0$/,
  privateClassA: /^10\.\d{1,3}\.\d{1,3}\.\d{1,3}$/,
  privateClassB: /^172\.(1[6-9]|2\d|3[0-1])\.\d{1,3}\.\d{1,3}$/,
  privateClassC: /^192\.168\.\d{1,3}\.\d{1,3}$/,
  loopbackIPv6: /^\[?::1\]?$/,
  localDomain: /\.local$/i,
  internalDomain: /\.internal$/i,
};

/**
 * Allowed URL protocols
 */
const ALLOWED_PROTOCOLS = new Set(['http:', 'https:']);

/**
 * Validate a URL and return the result
 *
 * @param url - URL string to validate
 * @param config - Optional validation configuration
 * @returns Validation result with normalized URL or error
 */
export function validateUrl(
  url: string,
  config: UrlValidatorConfig = {}
): UrlValidationResult {
  const {
    maxLength = DEFAULT_MAX_LENGTH,
    allowedDomains = [],
    blockedDomains = [],
    allowInternal = false,
  } = config;

  // Trim and check for empty
  const trimmedUrl = url?.trim() ?? '';
  if (!trimmedUrl) {
    return {
      valid: false,
      error: {
        code: 'INVALID_FORMAT',
        message: 'URL cannot be empty',
      },
    };
  }

  // Check length before processing
  if (trimmedUrl.length > maxLength) {
    return {
      valid: false,
      error: {
        code: 'URL_TOO_LONG',
        message: `URL exceeds maximum length of ${maxLength} characters`,
        details: { length: trimmedUrl.length, maxLength },
      },
    };
  }

  // Add protocol if missing
  let urlWithProtocol = trimmedUrl;
  if (!trimmedUrl.includes('://')) {
    urlWithProtocol = `https://${trimmedUrl}`;
  }

  // Parse URL
  let parsedUrl: URL;
  try {
    parsedUrl = new URL(urlWithProtocol);
  } catch {
    return {
      valid: false,
      error: {
        code: 'INVALID_FORMAT',
        message: 'Invalid URL format',
        details: { url: trimmedUrl },
      },
    };
  }

  // Check protocol
  const protocol = parsedUrl.protocol.toLowerCase();
  if (!ALLOWED_PROTOCOLS.has(protocol)) {
    return {
      valid: false,
      error: {
        code: 'INVALID_PROTOCOL',
        message: `Invalid protocol: ${protocol}. Only HTTP and HTTPS are allowed`,
        details: { protocol },
      },
    };
  }

  // Check for credentials in URL
  if (parsedUrl.username || parsedUrl.password) {
    return {
      valid: false,
      error: {
        code: 'CREDENTIALS_IN_URL',
        message: 'URLs with embedded credentials are not allowed',
      },
    };
  }

  // Get hostname for further checks
  const hostname = parsedUrl.hostname.toLowerCase();

  // Check for internal addresses (unless explicitly allowed)
  if (!allowInternal) {
    const internalCheck = isInternalAddress(hostname);
    if (internalCheck) {
      return {
        valid: false,
        error: {
          code: 'INTERNAL_ADDRESS',
          message: `Internal addresses are not allowed: ${hostname}`,
          details: { hostname, pattern: internalCheck },
        },
      };
    }
  }

  // Check blocked domains first (takes precedence)
  if (blockedDomains.length > 0) {
    for (const blockedDomain of blockedDomains) {
      if (domainMatches(hostname, blockedDomain.toLowerCase())) {
        return {
          valid: false,
          error: {
            code: 'DOMAIN_BLOCKED',
            message: `Domain is blocked: ${hostname}`,
            details: { hostname, blockedDomain },
          },
        };
      }
    }
  }

  // Check allowed domains (if specified)
  if (allowedDomains.length > 0) {
    const isAllowed = allowedDomains.some((allowedDomain) =>
      domainMatches(hostname, allowedDomain.toLowerCase())
    );
    if (!isAllowed) {
      return {
        valid: false,
        error: {
          code: 'DOMAIN_NOT_ALLOWED',
          message: `Domain is not in the allowed list: ${hostname}`,
          details: { hostname, allowedDomains },
        },
      };
    }
  }

  // Upgrade to HTTPS if HTTP
  if (protocol === 'http:') {
    parsedUrl.protocol = 'https:';
  }

  // Final length check after normalization
  const normalizedUrl = parsedUrl.href;
  if (normalizedUrl.length > maxLength) {
    return {
      valid: false,
      error: {
        code: 'URL_TOO_LONG',
        message: `Normalized URL exceeds maximum length of ${maxLength} characters`,
        details: { length: normalizedUrl.length, maxLength },
      },
    };
  }

  return {
    valid: true,
    url: normalizedUrl,
  };
}

/**
 * Check if a hostname is an internal/private address
 *
 * @param hostname - Hostname to check
 * @returns The matching pattern name if internal, null otherwise
 */
function isInternalAddress(hostname: string): string | null {
  for (const [name, pattern] of Object.entries(INTERNAL_PATTERNS)) {
    if (pattern.test(hostname)) {
      return name;
    }
  }
  return null;
}

/**
 * Check if a hostname matches a domain pattern (including subdomains)
 *
 * @param hostname - Full hostname to check
 * @param domain - Domain pattern to match against
 * @returns True if the hostname matches or is a subdomain
 */
export function domainMatches(hostname: string, domain: string): boolean {
  // Exact match
  if (hostname === domain) {
    return true;
  }
  // Subdomain match (hostname ends with .domain)
  if (hostname.endsWith(`.${domain}`)) {
    return true;
  }
  return false;
}

/**
 * URL Validator class for reusable validation with configuration
 */
export class UrlValidator {
  private config: UrlValidatorConfig;

  constructor(config: UrlValidatorConfig = {}) {
    this.config = config;
  }

  /**
   * Validate a URL using the configured settings
   */
  validate(url: string): UrlValidationResult {
    return validateUrl(url, this.config);
  }

  /**
   * Extract the domain (hostname) from a URL
   */
  extractDomain(url: string): string | null {
    try {
      const parsed = new URL(url);
      return parsed.hostname.toLowerCase();
    } catch {
      return null;
    }
  }

  /**
   * Check if a hostname matches a domain pattern
   */
  domainMatches(hostname: string, domain: string): boolean {
    return domainMatches(hostname, domain);
  }

  /**
   * Update configuration
   */
  updateConfig(config: Partial<UrlValidatorConfig>): void {
    this.config = { ...this.config, ...config };
  }

  /**
   * Get current configuration
   */
  getConfig(): UrlValidatorConfig {
    return { ...this.config };
  }
}

File: packages/agent-rs/crates/tron-tools/src/web/url_validator.rs
//! URL validation, blocked domains, and protocol upgrade.
//!
//! Validates URLs for safety (blocks private IPs, localhost, internal domains)
//! and auto-upgrades HTTP to HTTPS.

use regex::Regex;
use url::Url;

const MAX_URL_LENGTH: usize = 2000;

/// URL validation error codes.
#[derive(Debug, PartialEq, Eq)]
pub enum UrlError {
    /// URL is empty or invalid format.
    InvalidFormat(String),
    /// URL exceeds maximum length.
    TooLong,
    /// Protocol is not HTTP or HTTPS.
    InvalidProtocol(String),
    /// URL contains embedded credentials.
    CredentialsInUrl,
    /// URL points to a private/internal address.
    InternalAddress(String),
    /// Domain is in the blocked list.
    DomainBlocked(String),
    /// Domain is not in the allowed list.
    DomainNotAllowed(String),
}

impl std::fmt::Display for UrlError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::InvalidFormat(msg) => write!(f, "Invalid URL: {msg}"),
            Self::TooLong => write!(
                f,
                "URL exceeds maximum length of {MAX_URL_LENGTH} characters"
            ),
            Self::InvalidProtocol(p) => {
                write!(f, "Invalid protocol: {p} (only http/https allowed)")
            }
            Self::CredentialsInUrl => write!(f, "URL must not contain credentials"),
            Self::InternalAddress(host) => write!(f, "Internal/private address blocked: {host}"),
            Self::DomainBlocked(d) => write!(f, "Domain blocked: {d}"),
            Self::DomainNotAllowed(d) => write!(f, "Domain not in allowed list: {d}"),
        }
    }
}

/// Configuration for URL validation.
#[derive(Default)]
pub struct UrlValidatorConfig {
    /// Only allow these domains (empty = allow all).
    pub allowed_domains: Vec<String>,
    /// Block these domains.
    pub blocked_domains: Vec<String>,
}

/// Validate and normalize a URL.
///
/// Returns the validated HTTPS URL string, or an error.
pub fn validate_url(raw_url: &str, config: &UrlValidatorConfig) -> Result<String, UrlError> {
    let trimmed = raw_url.trim();
    if trimmed.is_empty() {
        return Err(UrlError::InvalidFormat("URL is empty".into()));
    }

    if trimmed.len() > MAX_URL_LENGTH {
        return Err(UrlError::TooLong);
    }

    let parsed = Url::parse(trimmed).map_err(|e| UrlError::InvalidFormat(e.to_string()))?;

    // Protocol check
    match parsed.scheme() {
        "http" | "https" => {}
        other => return Err(UrlError::InvalidProtocol(other.into())),
    }

    // Credentials check
    if !parsed.username().is_empty() || parsed.password().is_some() {
        return Err(UrlError::CredentialsInUrl);
    }

    // Host check
    let host = parsed
        .host_str()
        .ok_or_else(|| UrlError::InvalidFormat("no host in URL".into()))?;

    // Internal address check
    if is_internal_address(host) {
        return Err(UrlError::InternalAddress(host.into()));
    }

    // Domain filtering
    if !config.blocked_domains.is_empty() && domain_in_list(host, &config.blocked_domains) {
        return Err(UrlError::DomainBlocked(host.into()));
    }

    if !config.allowed_domains.is_empty() && !domain_in_list(host, &config.allowed_domains) {
        return Err(UrlError::DomainNotAllowed(host.into()));
    }

    // Auto-upgrade HTTP to HTTPS
    let mut result = parsed;
    if result.scheme() == "http" {
        let _ = result.set_scheme("https");
    }

    Ok(result.to_string())
}

/// Check if a hostname matches any domain in the list (subdomain-aware).
fn domain_in_list(host: &str, domains: &[String]) -> bool {
    let host_lower = host.to_lowercase();
    domains.iter().any(|d| {
        let d_lower = d.to_lowercase();
        host_lower == d_lower || host_lower.ends_with(&format!(".{d_lower}"))
    })
}

fn is_internal_address(host: &str) -> bool {
    let patterns = [
        r"^localhost$",
        r"^127\.\d{1,3}\.\d{1,3}\.\d{1,3}$",
        r"^0\.0\.0\.0$",
        r"^10\.\d{1,3}\.\d{1,3}\.\d{1,3}$",
        r"^172\.(1[6-9]|2\d|3[0-1])\.\d{1,3}\.\d{1,3}$",
        r"^192\.168\.\d{1,3}\.\d{1,3}$",
        r"^\[?::1\]?$",
        r"\.local$",
        r"\.internal$",
    ];
    let host_lower = host.to_lowercase();
    patterns
        .iter()
        .any(|p| Regex::new(p).is_ok_and(|re| re.is_match(&host_lower)))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn default_config() -> UrlValidatorConfig {
        UrlValidatorConfig::default()
    }

    #[test]
    fn valid_https_url() {
        let r = validate_url("https://example.com/page", &default_config());
        assert!(r.is_ok());
        assert!(r.unwrap().starts_with("https://"));
    }

    #[test]
    fn http_auto_upgraded_to_https() {
        let r = validate_url("http://example.com/page", &default_config()).unwrap();
        assert!(r.starts_with("https://"));
    }

    #[test]
    fn missing_protocol_returns_error() {
        let r = validate_url("example.com/page", &default_config());
        assert!(matches!(r, Err(UrlError::InvalidFormat(_))));
    }

    #[test]
    fn invalid_url_format() {
        let r = validate_url("not a url at all", &default_config());
        assert!(matches!(r, Err(UrlError::InvalidFormat(_))));
    }

    #[test]
    fn empty_url() {
        let r = validate_url("", &default_config());
        assert!(matches!(r, Err(UrlError::InvalidFormat(_))));
    }

    #[test]
    fn url_too_long() {
        let long = format!("https://example.com/{}", "a".repeat(2000));
        let r = validate_url(&long, &default_config());
        assert!(matches!(r, Err(UrlError::TooLong)));
    }

    #[test]
    fn blocked_localhost() {
        let r = validate_url("https://localhost/page", &default_config());
        assert!(matches!(r, Err(UrlError::InternalAddress(_))));
    }

    #[test]
    fn blocked_127_0_0_1() {
        let r = validate_url("https://127.0.0.1/page", &default_config());
        assert!(matches!(r, Err(UrlError::InternalAddress(_))));
    }

    #[test]
    fn blocked_private_192_168() {
        let r = validate_url("https://192.168.1.1/page", &default_config());
        assert!(matches!(r, Err(UrlError::InternalAddress(_))));
    }

    #[test]
    fn blocked_private_10_x() {
        let r = validate_url("https://10.0.0.1/page", &default_config());
        assert!(matches!(r, Err(UrlError::InternalAddress(_))));
    }

    #[test]
    fn allowed_domains_filter() {
        let config = UrlValidatorConfig {
            allowed_domains: vec!["example.com".into()],
            ..Default::default()
        };
        let r = validate_url("https://example.com/page", &config);
        assert!(r.is_ok());

        let r2 = validate_url("https://other.com/page", &config);
        assert!(matches!(r2, Err(UrlError::DomainNotAllowed(_))));
    }

    #[test]
    fn blocked_domains_filter() {
        let config = UrlValidatorConfig {
            blocked_domains: vec!["evil.com".into()],
            ..Default::default()
        };
        let r = validate_url("https://evil.com/page", &config);
        assert!(matches!(r, Err(UrlError::DomainBlocked(_))));

        let r2 = validate_url("https://good.com/page", &config);
        assert!(r2.is_ok());
    }
}
